File: marshitronica/scheduler.py
import time
import threading
import collections


class Scheduler(object):
    """Scheduler
    """
    def __init__(self):
        self.is_running = False
        self.start_time = None
        self.step_second = 0.0001
        self.queue = collections.OrderedDict()
        self.th = None
        self.bpm = 120

    def __run(self):
        while self.is_running:
            if len(self.queue):
                next = list(self.queue.keys())[0]
                if next < time.time() - self.start_time:
                    callables = self.queue.pop(next)
                    for collable in callables:
                        collable()
                    continue
            time.sleep(self.step_second)

    def start(self):
        """Start
        """
        # TODO: if started, raise Exception
        self.is_running = True
        self.start_time = time.time()
        self.th = threading.Thread(target=self.__run)
        self.th.start()

    def stop(self):
        """Stop
        """
        self.is_running = False
        self.start_time = None

    def time_to_beats(self, time):
        return time * self.bpm / 60

    def beats_to_time(self, beats):
        return beats * 60 / self.bpm

    @property
    def now(self):
        return time.time() - self.start_time

    @property
    def now_beats(self):
        return self.time_to_beats(self.now)

    def add_callable(self, beats, callable):
        t = self.beats_to_time(beats)

        if t < time.time() - self.start_time:
            return

        if t not in self.queue:
            self.queue[t] = [callable]
        else:
            self.queue.append(callable)

File: src/scheduler.cpp
// Copyright[2019] <marshi(masashi yoshikawa)>
#include "scheduler.h"
#include "unistd.h"


Scheduler::Scheduler() {
  return;
}

Scheduler::~Scheduler() {
  this->set_is_running(false);
  if (this->scheduling_thread.joinable()) {
    this->scheduling_thread.join();
  }
  return;
}

void Scheduler::schedule() {
  while (true) {
    this->is_running_mutex.lock();
    if (!this->is_running) {
      this->is_running_mutex.unlock();
      break;
    } else {
      this->is_running_mutex.unlock();
    }

    this->task_queue_mutex.lock();
    if (this->task_queue.size() > 0) {
      // TODO(marshi): no auto use
      auto next = this->task_queue.begin();
      double next_beats = next->first;
      std::vector<std::function<void(void)>> next_tasks = next->second;
      if (next_beats < this->now_beats()) {
        // TODO(marshi): threading
        for (std::function<void(void)> task : next_tasks) {
          task();
        }
        this->task_queue.erase(next_beats);
        this->task_queue_mutex.unlock();
        continue;
      }
    }
    this->task_queue_mutex.unlock();
    usleep(this->step_second * 1000000);
  }
}

void Scheduler::set_is_running(bool is_running) {
  this->is_running_mutex.lock();
  this->is_running = is_running;
  this->is_running_mutex.unlock();
}

void Scheduler::start() {
  this->set_is_running(true);
  this->start_time = std::chrono::system_clock::now();
  this->scheduling_thread = std::thread([this](){this->schedule();});
}

void Scheduler::stop() {
  this->set_is_running(false);
  this->scheduling_thread.join();
}

double Scheduler::seconds_to_beats(double seconds) {
  return seconds * this->bpm / 60;
}

double Scheduler::beats_to_seconds(double beats) {
  return beats * 60 / this->bpm;
}

double Scheduler::now_seconds() {
  std::chrono::duration<double> now_seconds_duration =
    std::chrono::system_clock::now() - this->start_time;
  return now_seconds_duration.count();
}

double Scheduler::now_beats() {
  return this->seconds_to_beats(this->now_seconds());
}

void Scheduler::add_task(double beats, std::function<void(void)> task) {
  if (beats < this->now_beats()) {
    return;
  }

  this->task_queue_mutex.lock();
  if (this->task_queue.find(beats) != this->task_queue.end()) {
    this->task_queue.at(beats).push_back(task);
  } else {
    this->task_queue.insert(
        std::make_pair(beats, std::vector<std::function<void(void)>>({task})));
  }
  this->task_queue_mutex.unlock();
}
